File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/features/users/createUser/CreateUserController.java
package br.com.deltaglobalbank.identity.features.users.createUser;

import java.util.UUID;

import br.com.deltaglobalbank.sharedauth.AuthenticatedPrincipal;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;

@Controller
public class CreateUserController {

    private final CreateUserUseCase createUserUseCase;

    public CreateUserController(CreateUserUseCase createUserUseCase) {
        this.createUserUseCase = createUserUseCase;
    }

    @PreAuthorize("hasRole('identity.admin')")
    @PostMapping("/admin/users")
    public ResponseEntity<CreateUserResponse> createUserWithImplicitTenantId(
        @AuthenticationPrincipal AuthenticatedPrincipal principal,
        @Valid @RequestBody CreateUserRequest request
    ) {
        CreateUserCommand command = new CreateUserCommand(
            principal.tenantId(),
            request.fullName(),
            request.email().value(),
            request.roleCodes(),
            principal.subject(),
            principal.roles()
        );

        CreateUserResponse response = createUserUseCase.execute(command);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @PreAuthorize("hasRole('platform.admin')")
    @PostMapping("/admin/tenants/{tenantId}/users")
    public ResponseEntity<CreateUserResponse> createUserWithExplicitTenantId(
        @AuthenticationPrincipal AuthenticatedPrincipal principal,
        @PathVariable UUID tenantId,
        @Valid @RequestBody CreateUserRequest request
    ) {
        CreateUserCommand command = new CreateUserCommand(
            tenantId,
            request.fullName(),
            request.email().value(),
            request.roleCodes(),
            principal.subject(),
            principal.roles()
        );
        CreateUserResponse response = createUserUseCase.execute(command);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/features/users/createUser/CreateUserController.kt
package br.com.deltaglobalbank.identity.features.users.createUser

import br.com.deltaglobalbank.sharedauth.AuthenticatedPrincipal
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import java.util.UUID

@Controller()
class CreateUserController(private val createUserUseCase: CreateUserUseCase) {

    @PreAuthorize("hasRole('identity.admin')")
    @PostMapping("/admin/users")
    fun createUserWithImplicitTenantId(
        @AuthenticationPrincipal principal: AuthenticatedPrincipal,
        @Valid @RequestBody request: CreateUserRequest
    ): ResponseEntity<CreateUserResponse> {
        val command = CreateUserCommand(
            tenantId = principal.tenantId,
            fullName = request.fullName,
            email = request.email.value,
            roleCodes = request.roleCodes,
            grantedBy = principal.subject,
            creatorRoles = principal.roles
        )

        val response = createUserUseCase.execute(command)
        return ResponseEntity.status(HttpStatus.CREATED).body(response)
    }

    @PreAuthorize("hasRole('platform.admin')")
    @PostMapping("/admin/tenants/{tenantId}/users")
    fun createUserWithExplicitTenantId(
        @AuthenticationPrincipal principal: AuthenticatedPrincipal,
        @PathVariable tenantId: UUID,
        @Valid @RequestBody request: CreateUserRequest
    ): ResponseEntity<CreateUserResponse> {
        val command = CreateUserCommand(
            tenantId = tenantId,
            fullName = request.fullName,
            email = request.email.value,
            roleCodes = request.roleCodes,
            grantedBy = principal.subject,
            creatorRoles = principal.roles
        )
        val response = createUserUseCase.execute(command)
        return ResponseEntity.status(HttpStatus.CREATED).body(response)
    }
}